pkg/controller/s3/s3bucket.go

```diff
@@ -18,249 +18,216 @@ package s3
-		bucket.Status.IAMUsername = s3.GenerateBucketUsername(bucket)
+	if aws.BoolValue(p.Spec.UseServiceAccount) {
+		s3Client, err := c.newClientFn(ctx, []byte{}, p.Spec.Region, awsclients.UsePodServiceAccount)
+		return &external{client: s3Client, kube: c.kube}, errors.Wrap(err, errCreates3Client)
 	}
 
-	// Get access keys for iam user
-	accessKeys, currentVersion, err := client.CreateUser(bucket.Status.IAMUsername, bucket)
-	if err != nil {
-		return r.fail(bucket, err)
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
 	}
 
-	// Set user policy version in status so we can detect policy drift
-	err = bucket.SetUserPolicyVersion(currentVersion)
-	if err != nil {
-		return r.fail(bucket, err)
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
 	}
 
-	if err := r.PublishConnection(ctx, bucket, managed.ConnectionDetails{
-		runtimev1alpha1.ResourceCredentialsSecretUserKey:     []byte(aws.StringValue(accessKeys.AccessKeyId)),
-		runtimev1alpha1.ResourceCredentialsSecretPasswordKey: []byte(aws.StringValue(accessKeys.SecretAccessKey)),
-		runtimev1alpha1.ResourceCredentialsSecretEndpointKey: []byte(bucket.Spec.Region),
-	}); err != nil {
-		return r.fail(bucket, err)
-	}
+	s3Client, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: s3Client, kube: c.kube}, errors.Wrap(err, errCreates3Client)
+}
 
-	// No longer creating, we're ready!
-	bucket.Status.SetConditions(runtimev1alpha1.Available())
-	resource.SetBindable(bucket)
-	return result, r.Update(ctx, bucket)
+type external struct {
+	client s3.Client
+	kube   client.Client
 }
 
-func (r *Reconciler) _sync(bucket *bucketv1alpha3.S3Bucket, client s3.Service) (reconcile.Result, error) {
-	if bucket.Status.IAMUsername == "" {
-		return r.fail(bucket, errors.New("username not set, .Status.IAMUsername"))
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.S3Bucket)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
 	}
-	bucketInfo, err := client.GetBucketInfo(bucket.Status.IAMUsername, bucket)
+
+	name := meta.GetExternalName(cr)
+
+	// Check if bucket in spec exists
+	_, err := e.client.HeadBucketRequest(&awsS3.HeadBucketInput{
+		Bucket: aws.String(name),
+	}).Send(ctx)
+
 	if err != nil {
-		return r.fail(bucket, err)
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(s3.IsErrorNotFound, err), errGetBucket)
 	}
 
-	if bucketInfo.Versioning != bucket.Spec.Versioning {
-		err := client.UpdateVersioning(bucket)
-		if err != nil {
-			return r.fail(bucket, err)
-		}
-	}
+	policy := e.getBucketPolicy(meta.GetExternalName(cr))
+
```

Comment:
It's a bit of personal preference, but I don't think empty lines help the reader understand the code better unless they separate steps that each consist of bulk of calls.